src/Uno.UI.RuntimeTests/Tests/Windows_UI_Xaml_Controls/Given_TreeView.cs

```diff
@@ -123,6 +122,89 @@ public async Task When_Open_Close_Twice_Grid()
 
 		Assert.AreEqual(child2, child2NodeAfter.DataContext);
 	}
+
+	[TestMethod]
+	public async Task When_Open_Close_Root_Twice_Keep_State()
+	{
+		var container = new Grid();
+		container.Width = 500;
+		container.Height = 500;
+
+		var SUT = new When_Open_Close_Twice_Grid();
+
+		container.Children.Add(SUT);
+
+		TestServices.WindowHelper.WindowContent = container;
+
+		var root = new MyNode();
+		root.Name = "root 4";
+		root.IsDirectory = true;
+		var child1 = new MyNode { Name = "Child 1", IsDirectory = true };
+		var child2 = new MyNode { Name = "Child 2", IsDirectory = true };
+		var child3 = new MyNode { Name = "Child 3", IsDirectory = true };
+		root.Children.Add(child1);
+		root.Children.Add(child2);
+		root.Children.Add(child3);
+
+		var child11 = new MyNode { Name = "Child 11" };
+		var child12 = new MyNode { Name = "Child 12" };
+		child1.Children.Add(child11);
+		child1.Children.Add(child12);
+
+		var child21 = new MyNode { Name = "Child 21" };
+		var child22 = new MyNode { Name = "Child 22" };
+		child2.Children.Add(child21);
+		child2.Children.Add(child22);
+
+		var child31 = new MyNode { Name = "Child 31" };
+		var child32 = new MyNode { Name = "Child 32" };
+		child3.Children.Add(child31);
+		child3.Children.Add(child32);
+
+		SUT.myTree.ItemsSource = new[] { root };
+		await TestServices.WindowHelper.WaitForIdle();
+
+		var rootNode = (TreeViewItem)SUT.myTree.ContainerFromItem(root);
+		rootNode.IsExpanded = true;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		var child2Node = (TreeViewItem)SUT.myTree.ContainerFromItem(child2);
+		child2Node.IsExpanded = true;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		rootNode.IsExpanded = false;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		rootNode.IsExpanded = true;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		// Force refresh
+		container.Width = 700;
+		container.Height = 700;
+
+		await TestServices.WindowHelper.WaitForIdle();
+
+		var child1Node = (TreeViewItem)SUT.myTree.ContainerFromItem(child1);
+		Assert.IsNotNull(child1Node);
+		Assert.AreEqual(child1, child1Node.DataContext);
+
+		rootNode.IsExpanded = false;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		rootNode.IsExpanded = true;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		var child3Node = (TreeViewItem)SUT.myTree.ContainerFromItem(child3);
+
+		// When index mappings are incorrect, the third node is expanded because
+		// of recursive items removal in the TreeView control.
+		Assert.IsFalse(child3Node.IsExpanded);
+		child3Node.IsExpanded = false;
+		await TestServices.WindowHelper.WaitForIdle();
+
+		child3Node.IsExpanded = true;
+		await TestServices.WindowHelper.WaitForIdle();
```

Comment:
arent we done asserting from here on?

Reply:
We should be. I tried to find a possible validation but could not find any. It's all internal to the ListView, though maybe I could enumerate all possible children an count valid items. I'll change it :)